doc/developer/design/20230418_persist_state_pubsub.md

```diff
@@ -0,0 +1,348 @@
+    materialize.cloud/envd-generation: <envd-generation>
+  type: ClusterIP
+  clusterIP: None
+  ports:
+  - name: pubsub
+    port: 6879
+    protocol: TCP
+    targetPort: 6879
+```
+
+`envd` will be informed of this service name via command line argument, e.g. `--persist-pubsub-service-name persist-pubsub-1`
+and pass it along to its `clusterd` param when set.
+
+We introduce this new headless service, rather than reusing the existing `environmentd` Service for a few reasons:
+
+* The `environmentd` Service is public / external-facing, while the pubsub port should be strictly internal
+* To decouple the notion that `environmentd` is necessarily responsible for providing the PubSub server. In the
+future we may wish to [replace it with a different implementation](#external-pubsubmessage-bus), or route the Service to
+point towards different pods.
+
+We will create the Service per `envd` generation, so that `clusterd` will always be routed to the `envd` matching their
+generation, and that multiple generations may coexist.
+
+# Rollout
+[rollout]: #rollout
+
+We will roll out this change in phases behind a feature flag. The change is seen purely as an optimization opportunity,
+and should be safe to shut off at any time. We will aim to ensure that any errors handled due to Persist PubSub cannot
+crash or halt the process.
+
+We anticipate message volume to be on the order of 1-1.5 per second per shard, with each message in the 100 bytes-1KiB
+range. Given the scale of our usage today, we would expect `environmentd` to be able to broadcast this data and message
+volume comfortably with minimal impact on performance.
+
+That said, we'll want to keep an eye on how much time `environmentd` spends broadcasting diffs, as it will be a new
+workload for the process. If `environmentd` CPU usage spikes too high (e.g. >0.5-1 CPU), we will reconsider the approach
+entirely. Early prototyping has indicated a very small rise in CPU (0.05-0.1 range for a heavy workload).
+
+We anticipate `clusterd` to be negligibly impacted by this added RPC traffic, as each `clusterd` would only need to
+push/receive updates for the shards it is reading and writing, work is that is very similar to what it is doing today.
+
+Our existing Persist metrics dashboard, in addition to the new metrics outlined below, should be sufficient to monitor
+the change and understand its impact.
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+We will introduce several metrics to understand the behavior of the change:
+
+* Number of updates pushed
+* Number of updates received that apply cleanly & uncleanly
+* How frequently `Listen`s need to fallback to polling
+* End-to-end latency of appending data to observing it
+* Timings of `environmentd`'s broadcasting / load
+
+CI will be used to ensure we do not introduce novel crashes or halts, but we anticipate the most interesting data to
```

Comment:
@philip-stoev @def- would love any ideas you have about testing this change. We'll be writing some unit tests to cover the basic logic, but since this change will be introducing some new networking to `environmentd` and `clusterd`, in addition to giving us a new path for applying updates to persist state, perhaps we want to stress it more? I'm particularly concerned about ensuring the clients reconnect if the server restarts / making sure the code does not panic / that processes spin up successfully if the feature is disabled or params are incorrect

Reply:
Let us wrap our heads around this and will get back to you. At the very least, we need to start using multi-process sources/sinks in the CI (as opposed to multi-worker). 

re restarts/reconnects, we will come up with a separate mzcompose workflows as needed.

Reply:
https://github.com/MaterializeInc/materialize/issues/19246 . Please add any comments about testing or QA to the issue.

Reply:
Sounds great, thank you @philip-stoev!